engine/src/test/scala/cromwell/engine/workflow/WorkflowDockerLookupActorSpec.scala

```diff
@@ -0,0 +1,391 @@
+    lookupActor ! yesterdaysDockerRequest
+
+    dockerHashingActor.expectMsg(latestDockerRequest)
+    dockerHashingActor.expectMsg(yesterdaysDockerRequest)
+    val latestSuccessResponse = DockerHashSuccessResponse(DockerHashResult("md5", "AAAAA"), latestDockerRequest)
+    val yesterdaysSuccessResponse = DockerHashSuccessResponse(DockerHashResult("md5", "BBBBB"), yesterdaysDockerRequest)
+    dockerHashingActor.reply(latestSuccessResponse)
+    dockerHashingActor.reply(yesterdaysSuccessResponse)
+
+    val results = Set(1, 2) map { _ =>
+      expectMsgPF(2 seconds) {
+        case result: DockerHashSuccessResponse => result
+      }
+    }
+
+    results should equal(Set(latestSuccessResponse, yesterdaysSuccessResponse))
+  }
+
+  it should "handle hash write errors appropriately" in {
+    val dockerValue = "ubuntu:latest"
+    val dockerId = DockerImageIdentifier.fromString(dockerValue).get.asInstanceOf[DockerImageIdentifierWithoutHash]
+
+    val workflowId = WorkflowId.randomId()
+    val dockerHashingActor = TestProbe()
+
+    val databaseConfig = ConfigFactory.load.getConfig("database")
+
+    var numWrites = 0
+
+    val databaseInterface = new SlickDatabase(databaseConfig) {
+
+      override def queryDockerHashStoreEntries(workflowExecutionUuid: String)
+                                              (implicit ec: ExecutionContext): Future[Seq[DockerHashStoreEntry]] =
+        throw new RuntimeException("should not be called, this workflow is not restarting")
+
+      override def addDockerHashStoreEntry(dockerHashStoreEntry: DockerHashStoreEntry)(implicit ec: ExecutionContext): Future[Unit] = {
+        numWrites = numWrites + 1
+        if (numWrites == 1) Future.failed(new RuntimeException("Fake exception from a test.")) else Future.successful(())
+      }
+    }.initialized
+
+    val lookupActor = TestActorRef(WorkflowDockerLookupActor.props(workflowId, dockerHashingActor.ref, StartNewWorkflow, databaseInterface))
+    val request = DockerHashRequest(dockerId)
+    lookupActor ! request
+    val response = DockerHashSuccessResponse(DockerHashResult("md5", "AAAAAAAA"), request)
+
+    // The WorkflowDockerLookupActor should not have the hash for this tag yet and will need to query the dockerHashingActor.
+    dockerHashingActor.expectMsg(request)
+    dockerHashingActor.reply(response)
+    // The WorkflowDockerLookupActor is going to fail when it tries to write to that broken DB.
+    expectMsgPF(2 seconds) {
+      case fail: WorkflowDockerLookupFailure =>
+    }
+    numWrites should equal(1)
+
+    lookupActor ! request
+    // The WorkflowDockerLookupActor will query the dockerHashingActor again.
+    dockerHashingActor.expectMsg(request)
+    dockerHashingActor.reply(response)
+    // The WorkflowDockerLookupActor should forward the success message to this actor.
+    expectMsg(response)
+    numWrites should equal(2)
+  }
+
+  it should "emit a terminal failure message if failing to read hashes on restart" in {
```

Comment:
For restart, all hashes are read in a bolus up front before any work begins on the workflow.  If we can't do even that it was decided we should just stop rather than start using hashes that might be inconsistent with what was previously used.